### PR TITLE
Add globallyDisableIrqLoadBalancing flag to Performance Profile

### PR DESCRIPTION
**Pull request by marcel-apf**
v5:
 - Solved a bug in the E2E test limiting it to 64 CPUs machines (@fromanirh)
 - Rebased on master

v4:
 - Add an E2E testing the GloballyDisableIrqLoadBalancing Spec field

v3:
 - Fix API conversion test bug found by CI

v2:
- Rebased on master
- Add a comment on why we are making the change (@fromanirh )
- Solved a bug and updated the template with a new logic

### CODE REVIEW

File: build/assets/tuned/openshift-node-performance
Comment (fromanirh) [inline build/assets/tuned/openshift-node-performance:50]: let's loop in @yanirq

File: build/assets/tuned/openshift-node-performance
Comment (fromanirh) [inline build/assets/tuned/openshift-node-performance:49]: do we need to have this if GloballyDisableIrqLoadBalancing is false? meaning: could the following work?
```
{{if .GloballyDisableIrqLoadBalancing}}
[irqbalance]
banned_cpus=${isolated_cores}
{{end}}
```

**Review by fromanirh — COMMENTED** [review]
at glance looks nice, will have a deeper review later

File: build/assets/tuned/openshift-node-performance
Comment (marcel-apf) [reply build/assets/tuned/openshift-node-performance:49]: The parent tuned profile has:
```
[irqbalance]
banned_cpus=${isolated_cores}
```
I wanted to override that. [I will double check]

File: build/assets/tuned/openshift-node-performance
Comment (marcel-apf) [reply build/assets/tuned/openshift-node-performance:50]: Good idea

File: build/assets/tuned/openshift-node-performance
Comment (fromanirh) [reply build/assets/tuned/openshift-node-performance:49]: OK, if you need to override it it's fine like it is.

Comment (marcel-apf) [reply build/assets/tuned/openshift-node-performance:49]: Now I am too (depends on the implementation of tuned). We will know in about 1 hour

Comment (yanirq) [reply build/assets/tuned/openshift-node-performance:49]: I would dig a bit more in https://github.com/openshift/cluster-node-tuning-operator/blob/master/assets/tuned/daemon/tuned/plugins/plugin_irqbalance.py to see the implications of an empty string and also see what you actually set on the system

Comment (yanirq) [reply build/assets/tuned/openshift-node-performance:49]: we had a test once checking IRQBALANCE_BANNED_CPUS value in the irqbalance config file when we had initramfs generated. maybe its time to bring this test back.

Comment (marcel-apf) [reply build/assets/tuned/openshift-node-performance:49]: @yanirq I actually tested live both cases and on empty string the cpu mask was 0 as expected
Regarding the mentioned test, I have something in mind, but  a cheat-sheet is always welcomed :) , do you remember the file or the test name? I will dig anyway.

Comment (marcel-apf) [reply build/assets/tuned/openshift-node-performance:49]: I've gone fast over the tuned code and it should be ok.
I will triple-check anyway, better safe...

**Comment by marcel-apf** [comment]
Several tests failed, looks like a glitch in CI (system too slow?)
Worth immediate retest
/retest

File: build/assets/tuned/openshift-node-performance
Comment (marcel-apf) [reply build/assets/tuned/openshift-node-performance:49]: @fromanirh it took me a lot of time, but at least tuned started to work (MCO still can't install the rt kernel)
To answer your question, as I thought, if we don't add  banned_cpus="" the  isolated CPU set remains masked.
The reason is we inherit the cpu-partitioning bits, so we need to override banned_cpus with an empty value.
The other way around would work, e.g. setting banned_cpus only to empty string when needed, but I think a verbose mode is desired

**Comment by marcel-apf** [comment]
Since all the errors don't make sense, and it seems the system was slow all day long, last try today
/retest

File: build/assets/tuned/openshift-node-performance
Comment (fromanirh) [reply build/assets/tuned/openshift-node-performance:49]: thanks for the extra check @marcel-apf . I'm fine with your approach. I just think this
```
f we don't add banned_cpus="" the isolated CPU set remains masked.
The reason is we inherit the cpu-partitioning bits, so we need to override banned_cpus with an empty value.
```
should be added as code comment for future reference.

**Comment by marcel-apf** [comment]
Another day, another conclusion. The last failure is actually ok.
The reason is I started a new section [irqbalance] in the middle of [sysctl] one...

File: build/assets/tuned/openshift-node-performance
Comment (marcel-apf) [reply build/assets/tuned/openshift-node-performance]: I changed the template to:
```{{if not .GloballyDisableIrqLoadBalancing}}
[irqbalance]
# Override the value set by cpu-partitioning with an empty one
banned_cpus=""
{{end}}
```
The reason is we don't know what exactly the parent use as banned cpus, they may update the formula in the future.
What we do know is we don't want to bann the CPUs for our case

File: functests/utils/nodes/nodes.go
Comment (fromanirh) [inline functests/utils/nodes/nodes.go:190]: so this is a bitmask, right?
this worries me a bit because it suggests it can't hold more than 64 CPUs :\

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [inline functests/2_performance_update/updating_profile.go:93]: it's a test, so we don't care too much about the traffic. Can't we just update back with the original profile and save some complexity?

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [inline functests/2_performance_update/updating_profile.go:101]: do we need to wait here?

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [inline functests/2_performance_update/updating_profile.go:92]: I think we need to wait here, don't we?

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [inline functests/2_performance_update/updating_profile.go:104]: with the added waits(), we can use the same loop as line 74 above, which is nice for consistency

**Review by fromanirh — COMMENTED** [review]
my biggest concern is to wait after we restore the original profile, everything else which caught my eye is about possible improvements that can wait

File: functests/utils/nodes/nodes.go
Comment (marcel-apf) [reply functests/utils/nodes/nodes.go:190]: indeed, however the default /etc/sysconfig/irqbalance value suggests a 64bit field. I will dig more to see what is going on, I will increase the field space if necessary. (which probably is)

File: functests/2_performance_update/updating_profile.go
Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:93]: The snippet is taken from the same file: https://github.com/openshift-kni/performance-addon-operators/blob/master/functests/2_performance_update/updating_profile.go#L264

I really don't care one way or another, but should we be consistent?

File: functests/2_performance_update/updating_profile.go
Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:101]: We updated the profile. Now the tuned profile will be updated and the node will be updated. It happens async, right?

Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:101]: Following the other snippet I would say no. The update is very "fast". If we do wait, we should wait on both, right?

File: functests/2_performance_update/updating_profile.go
Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:92]: Not really, we wait once for the node to have the right settings. Anything in between does not worth the wait.

File: functests/2_performance_update/updating_profile.go
Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:104]: What do you mean? 
The first loop doesn't have to wait for anything, this one does.
If you have a nicer way to code it, please point me to it, I will gladly improve (I am not mastering the Ginko stuff yet)

File: functests/utils/nodes/nodes.go
Comment (fromanirh) [reply functests/utils/nodes/nodes.go:190]: IMO the best option would be the list format of cpuset, like linux does (https://man7.org/linux/man-pages/man7/cpuset.7.html#FORMATS)

Comment (marcel-apf) [reply functests/utils/nodes/nodes.go:190]: I Will follow, thanks!

Comment (fromanirh) [reply functests/utils/nodes/nodes.go:190]: sure, even though it seems for irqbalance we are out of luck and stuck with the bitmask :\
```
#
# IRQBALANCE_BANNED_CPUS
# 64 bit bitmask which allows you to indicate which cpu's should
# be skipped when reblancing irqs. Cpu numbers which have their
# corresponding bits set to one in this mask will not have any
# irq's assigned to them on rebalance
#
#IRQBALANCE_BANNED_CPUS=
```

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [reply functests/2_performance_update/updating_profile.go:92]: ok

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [reply functests/2_performance_update/updating_profile.go:93]: let's keep it consistent and change everything in one go.

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [reply functests/2_performance_update/updating_profile.go:101]: ok, just checking

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [reply functests/2_performance_update/updating_profile.go:104]: ok, we can keep as it is.

File: functests/utils/nodes/nodes.go
Comment (fromanirh) [reply functests/utils/nodes/nodes.go:190]: OK, I reviewed the irqbalance code and on their side they can handle up until 4096 cpus encoded as bitmask. So irqbalance is fine. The potential issue is on our side (test) if we represent the bitmap using native types, so we are limited by (u)int64.

Comment (marcel-apf) [reply functests/utils/nodes/nodes.go:190]: Right, also there is a parsing issue. It turns out they delimit each 32 CPUs with a comma, something like 00000000,00000000,....
Another thing that caught my mind is:
```
From Red Hat Enterprise Linux 7.2, the irqbalance tool automatically avoids IRQs on CPU cores isolated via the isolcpus= kernel parameter if IRQBALANCE_BANNED_CPUS is not set in the /etc/sysconfig/irqbalance file.
```
See https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_for_real_time/7/html/tuning_guide/interrupt_and_process_binding


That means what we are doing actually makes sense, setting the mask to "nothing", which is different from not setting it at all, in which case the irq load balancing will actually be disabled for the isolated cpus.

File: functests/2_performance_update/updating_profile.go
Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:93]: agreed, we need a "cleaning" PR.

File: functests/2_performance_update/updating_profile.go
Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:104]: Please elaborate more, I want to at lest know it. Can you add more context to the "Wait()" ?
Also we should have conventions.

File: functests/utils/nodes/nodes.go
Comment (marcel-apf) [reply functests/utils/nodes/nodes.go:190]: And regarding the CPU mask set, you are perfectly write. I didn't even know about it, thanks! I am going to see how it is done in go.

File: functests/2_performance_update/updating_profile.go
Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:101]: Once we re-write, we can add the wait as a good practice though.

File: functests/utils/nodes/nodes.go
Comment (marcel-apf) [reply functests/utils/nodes/nodes.go:190]: Turns out we are already using kubelet's (!!) cpuset.CPUSet type. Nice, I will use it.

File: functests/2_performance_update/updating_profile.go
Comment (fromanirh) [reply functests/2_performance_update/updating_profile.go:104]: Setting up conventions (but most notably enforce them) would be interesting.
Thing is the usage of Eventually() here suggests that changes to the cluster made in the previous steps of the tests (e.g. change the profile) may take some time to be accepted, and then the state of the system will Eventually() be what's expected - otherwise the test will fail.
That works.
Another way is to wait explicitely for the changes to be accepted by the cluster, hence after each stage is performed the state of the cluster is known (or otherwise a previous expectation failed, letting the test fail). Once the state of the cluster is known, there is often no need to use Eventually(), which may hide bugs.
But again is not a clear cut or an obvious choice: it depends on how the test is written, and sometimes depends also on other factos (e.g. are we trying to cut down the test execution time?)

Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:106]: Thanks for the explanation, so basically if we change the profile we should test that the PAO updated the profile and NTO also updated the tuned config AND finished to update the node. I think it will be doable if we'll have all this info in the PAO status section. Then we can simply wait for the status to be updated.  This is what you have in mind?

Comment (fromanirh) [reply functests/2_performance_update/updating_profile.go:106]: pretty much, yes. We can wait directly on MCP/NTO statuses if required or simpler. Anyway, no big deal for this PR.

**Review by fromanirh — APPROVED** [review]
/lgtm
/hold
nice!

letting other reviewers chime in

**Comment by marcel-apf** [comment]
```[1mSTEP[0m: Creating the PerformanceProfile
[1mSTEP[0m: Getting MCP for profile
W1104 14:56:55.547801   15587 config.go:57] A PerformanceProfile with name performance already exists! If created externally, tests might have unexpected behaviour
[1mSTEP[0m: MCP is already unpaused
[1mSTEP[0m: Waiting for the MCP to pick the PerformanceProfile's MC
[1mSTEP[0m: Waiting for MCP being updated
I1104 14:56:55.575281   15587 mcps.go:133] MCP is targeting 1 node(s)

[91m[1m• Failure [1800.050 seconds][0m```

Slow CI? I will wait 1 hour and try again

File: functests/1_performance/performance.go
Comment (yanirq) [inline functests/1_performance/performance.go:678]: must be set as in must have a value ?  can't it be nil ?  its defined as omitempty : 
`json:"globallyDisableIrqLoadBalancing,omitempty"`

Comment (marcel-apf) [reply functests/1_performance/performance.go:678]: Yes, is on purpose, when converting a v1 profile to v2 we must set the flag to true on order to keep the irq load balancing disabled as in v1 profile.

File: functests/2_performance_update/updating_profile.go
Comment (dshchedr) [inline functests/2_performance_update/updating_profile.go:60]: just curious, is it possible to make this test as part of common "performance profile updating" test?
https://github.com/openshift-kni/performance-addon-operators/blob/dfe203e14f050dda0f8f39dda24b567db6b91782/functests/2_performance_update/updating_profile.go#L123

It may help us to reduce wasting time on "updating / reverting back" node tunnings.

**Comment by marcel-apf** [comment]
slow CI, trying again
/retest

File: functests/2_performance_update/updating_profile.go
Comment (marcel-apf) [reply functests/2_performance_update/updating_profile.go:60]: I thought about it, but is a very fast test, is only updating the tuned profile, all the test takes a few seconds.
Also the test has checks before and after the update, it doesn't feel right to check the business logic in "Before All"

**Comment by marcel-apf** [comment]
/unhold
we have a second lgtm

**Comment by marcel-apf** [comment]
/retest
Ci bug fixed?

**Comment by fromanirh** [comment]
good news: CI seems to be healed
bad news: a (trivial) rebase is needed :\

**Comment by marcel-apf** [comment]
> good news: CI seems to be healed
> bad news: a (trivial) rebase is needed :\

done

**Pull request merged**
